File: practice/data_structures/linked_lists/reverse_doubly_linked_list/reverse_doubly_linked_list.py
"""
 Reverse a doubly linked list
 head could be None as well for empty list
 Node is defined as
 
 class Node(object):
 
   def __init__(self, data=None, next_node=None, prev_node = None):
       self.data = data
       self.next = next_node
       self.prev = prev_node

 return the head node of the updated list 
"""

def Reverse(i_head):
    if not i_head:
        return i_head
    while i_head.next:
        cp = i_head.next
        i_head.prev, i_head.next = i_head.next, i_head.prev
        i_head = cp
    i_head.prev, i_head.next = i_head.next, i_head.prev
    return i_head

File: practice/data_structures/linked_lists/reverse_doubly_linked_list/reverse_doubly_linked_list.cpp
/*
   Reverse a doubly linked list, input list may also be empty
   Node is defined as
   struct Node
   {
     int data;
     Node *next;
     Node *prev;
   }
*/

Node * Reverse(Node * i_head)
{
    if (!i_head)
    {
        return i_head;
    }
    
    while (i_head->next)
    {
        Node * node = i_head->next;
        swap(i_head->prev, i_head->next);
        i_head = node;
    }
    swap(i_head->prev, i_head->next);
    
    return i_head;
}

